### PR TITLE
Record a new sound from the mic

### PR DESCRIPTION
**Pull request by paulkaplan**
### Resolves

_What Github issue does this resolve (please include link)?_
Fixes https://github.com/LLK/scratch-gui/issues/296

### Proposed Changes

_Describe what this Pull Request does_
Adds a new microphone recording flow:
mic-full-flow

/cc @carljbowman @thisandagain

### CODE REVIEW

File: src/lib/audio/audio-recorder.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        onUpdate(rms);
+    };
+
+    requestAnimationFrame(update);
+
+    // Wire everything together, ending in the destination
+    this.mediaStreamSource.connect(this.sourceNode);
+    this.sourceNode.connect(this.analyserNode);
+    this.analyserNode.connect(this.scriptProcessorNode);
+    this.scriptProcessorNode.connect(this.audioContext.destination);
```
Comment (ericrosenbaum) [inline src/lib/audio/audio-recorder.js]: This line is unnecessary as far as I know!

Comment (paulkaplan) [inline src/lib/audio/audio-recorder.js]: (addressed offline, for posterity) it looks like it is needed.

**Comment by paulkaplan** [comment]
I'm wondering if the `lib/audio` code should really live in a separate repo (perhaps scratch-audio?) At first I thought the functionality was completely separate from scratch-audio, but since seeing @ericrosenbaums work recently, there is definitely a lot of overlap (like sampling RMS from an `analyserNode` and playing sound buffers). It would also keep the GUI as just front-end components. @ericrosenbaum what do you think about that? Is there a way we could use the `audioEngine` for the visualization / recording / playback I'm doing here?

**Comment by mrjacobbloom** [comment]
I'd lump the audio editor in with the paint editor, which seems big enough to live in its own repo. Just my 2 cents

File: src/containers/record-modal.jsx [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                    md5: `${md5}.wav`,
+                    name: `recording ${this.props.vm.editingTarget.sprite.sounds.length}`
+                };
+
+                // Load the encoded .wav into the storage cache
+                const storage = this.props.vm.runtime.storage;
+                storage.builtinHelper.cache(
+                    storage.AssetType.Sound,
+                    storage.DataFormat.WAV,
+                    new Uint8Array(wavBuffer), // @todo ask cwf about the need for this?
```
Comment (rschamp) [inline src/containers/record-modal.jsx]: @cwillisf There's a question here for you ^

File: src/containers/record-modal.jsx [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    handleStopPlaying () {
+        this.setState({playing: false});
+    }
+    handleBack () {
+        this.setState({playing: false, samples: null});
+    }
+    handleSubmit () {
+        this.setState({encoding: true});
+
+        // Allow enough time for UI refresh before starting encoding
```
Comment (rschamp) [inline src/containers/record-modal.jsx]: What's the refresh we're waiting for? Can this be scheduled more accurately than with a timeout?

File: src/lib/audio/audio-buffer-player.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+const SharedAudioContext = require('./shared-audio-context.js');
+
+const AudioBufferPlayer = function (samples) {
```
Comment (rschamp) [inline src/lib/audio/audio-buffer-player.js]: Is there a reason this doesn't use the new class syntax?

File: src/lib/audio/audio-recorder.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
+const SharedAudioContext = require('./shared-audio-context.js');
+
+const AudioRecorder = function () {
```
Comment (rschamp) [inline src/lib/audio/audio-recorder.js]: Same question here re: class syntax.

File: src/containers/record-modal.jsx (hunk 3 above)
Comment (paulkaplan) [inline src/containers/record-modal.jsx]: So my hope was that if the wav encoding took a bit of time, we would have enough time to switch this to say `Loading...` before starting the encoding process. Does putting this in a callback of `setState` do that?

Comment (rschamp) [inline src/containers/record-modal.jsx]: Oh, I see. Yeah, as far as I know, the callback works for that purpose. I think it's called after the re-render. `componentDidUpdate` is the recommended way, but that doesn't seem right for a handler like this.

File: src/components/record-modal/record-modal.css [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+.modal-overlay {
+    position: fixed;
+    top: 0;
+    left: 0;
+    right: 0;
+    bottom: 0;
+    z-index: 1000;
+    background-color: rgba(94, 102, 125, .95);
+}
```
Comment (paulkaplan) [inline src/components/record-modal/record-modal.css]: Yeah we no longer have a generic small modal component (the large modal component is now full-screen and uses a text filter in the nav bar, so isn't general enough anymore). Should probably be refactored.

**Comment by rschamp** [comment]
We should probably resolve the way to access the modal before merging.

**Comment by paulkaplan** [comment]
Almost forgot about that ray. I added that and also refactored the way the modal worked so that it could be used by the three modal-y things (prompt, recorder and library). Paired with Carl on the add buttons. There are some unusual border-related edge cases, but we aren't concerned about them. 

image

**Pull request merged**